File: server/src/intTest/java/org/tctalent/server/repository/db/OccupationSpecificationIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedOccupation;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.jpa.domain.Specification;
import org.tctalent.server.model.db.Occupation;
import org.tctalent.server.model.db.Status;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;
import org.tctalent.server.request.occupation.SearchOccupationRequest;

public class OccupationSpecificationIntTest extends BaseDBIntegrationTest {

  @Autowired
  private OccupationRepository repo;
  private Occupation occupation;
  private Specification<Occupation> spec;
  private SearchOccupationRequest request;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());
    occupation = getSavedOccupation(repo);

    request = new SearchOccupationRequest();
  }

  @Test
  public void testKeyword() {
    request.setKeyword(occupation.getName());
    spec = OccupationSpecification.buildSearchQuery(request);
    List<Occupation> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    assertEquals(1, results.size());
    assertEquals(occupation.getId(), results.getFirst().getId());
  }


  @Test
  public void testKeywordFail() {
    request.setKeyword("NOTHING");
    spec = OccupationSpecification.buildSearchQuery(request);
    List<Occupation> results = repo.findAll(spec);
    assertNotNull(results);
    assertTrue(results.isEmpty());
  }

  @Test
  public void testWithStatus() {
    request.setStatus(Status.active);
    spec = OccupationSpecification.buildSearchQuery(request);
    List<Occupation> results = repo.findAll(spec);
    assertNotNull(results);
    assertFalse(results.isEmpty());
    List<Long> ids = results.stream().map(Occupation::getId).toList();
    assertTrue(ids.contains(occupation.getId()));
  }

  @Test
  public void testWithStatusFail() {
    request.setStatus(Status.deleted);
    spec = OccupationSpecification.buildSearchQuery(request);
    List<Occupation> results = repo.findAll(spec);
    assertNotNull(results);
    assertTrue(results.isEmpty());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/OccupationSpecificationIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.tctalent.server.model.db.Occupation
import org.tctalent.server.model.db.Status
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest
import org.tctalent.server.repository.db.integrationhelp.getSavedOccupation
import org.tctalent.server.request.occupation.SearchOccupationRequest

class OccupationSpecificationIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: OccupationRepository
  private lateinit var occupation: Occupation

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    occupation = getSavedOccupation(repo)
  }

  @Test
  fun `test keyword`() {
    val request = SearchOccupationRequest().apply { keyword = occupation.name }
    val spec = OccupationSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue { result.isNotEmpty() }
    assertEquals(1, result.size)
    assertEquals(occupation.id, result.first().id)
  }

  @Test
  fun `test keyword fail`() {
    val request = SearchOccupationRequest().apply { keyword = "NOTHING" }
    val spec = OccupationSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue { result.isEmpty() }
  }

  @Test
  fun `test build search query with status`() {
    val request = SearchOccupationRequest().apply { status = Status.active }
    val spec = OccupationSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isNotEmpty())
    val ids = result.map { it.id }
    assertTrue { ids.contains(occupation.id) }
  }

  @Test
  fun `test build search query with status fail`() {
    val request = SearchOccupationRequest().apply { status = Status.deleted }
    val spec = OccupationSpecification.buildSearchQuery(request)
    val result = repo.findAll(spec)
    assertNotNull(result)
    assertTrue(result.isEmpty())
  }
}
